docs/docs/optimizing-performance.md

```diff
@@ -0,0 +1,118 @@
+---
+id: optimizing-performance
+title: Optimizing Performance
+permalink: docs/optimizing-performance.html
+---
+
+Internally, React uses several clever techniques to minimize the number of costly DOM operations required to update the UI. For many applications, using React will lead to a fast user interface without doing much work to specifically optimize for performance. Nevertheless, there are several ways you can speed up your React application.
+
+## Use The Production Build
+
+If you're benchmarking or experiencing performance problems in your React apps, make sure you're testing with the [minified production build](/react/downloads.html). The development build includes extra warnings that are helpful when building your apps, but it is slower due to the extra bookkeeping it does.
+
+## Avoid Reconciling the DOM
+
+React makes use of a "virtual DOM", which is a descriptor of a DOM subtree rendered in the browser. This parallel representation allows React to avoid creating DOM nodes and accessing existing ones, which is slower than operations on JavaScript objects. When a component's props or state change, React decides whether an actual DOM update is necessary by constructing a new virtual DOM and comparing it to the old one. When they are not equal, React will update the DOM, which can be a slow operation.
+
+In some cases, your component can speed all of this up by overriding the lifecycle function `shouldComponentUpdate`, which is triggered before the re-rendering process starts. The default implementation of this function returns `true`, leaving React to perform the update:
+
+```javascript
+shouldComponentUpdate(nextProps, nextState) {
+  return true;
+}
+```
+
+If you know that in some situations your component doesn't need to update, you can return `false` from `shouldComponentUpdate` instead, to skip the slow DOM update. For example, if the only way your component ever changes is when the `state.count` variable changes, you could write:
+
+```javascript
+shouldComponentUpdate(nextProps, nextState) {
+  return this.state.count !== nextState.count;
+}
+```
+
+Keep in mind that React will invoke this function pretty often, so the implementation has to be fast.
+
+## shouldComponentUpdate In Action
+
+Here's a subtree of components. For each one, `SCU` indicates what `shouldComponentUpdate` returned, and `vDOMEq` indicates whether the virtual DOMs were equivalent. Finally, the circle's color indicates whether the component had to be reconciled or not.
+
+<figure><img src="/react/img/docs/should-component-update.png" /></figure>
+
+Since `shouldComponentUpdate` returned `false` for the subtree rooted at C2, React did not generate a new virtual DOM for C2, and React didn't even have to invoke `shouldComponentUpdate` on C4 and C5.
```

Comment:
Since `shouldComponentUpdate` returned `false` for the subtree rooted at C2, React did not attempt to render C2, and thus didn't even have to invoke `shouldComponentUpdate` on C4 and C5.